### PR TITLE
Parallel snapshot

### PR DESCRIPTION
**Pull request by hemalathagajendran**
Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature Enhancement
- [ ] Test Automation 
- [ ] Code Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Community Operator listing 
- [ ] Other (please describe): 

## What is the current behavior?

- cg volume creation, deletion is not serialised

## What is the new behavior?

- cg volume creation, deletion is serialised. If cg snapshot creation is going on  and shallow copy creation/deletion is triggered with same target path then it would fail as the target path would be locked by CreateSnapshot and same would happen for vice versa as well
- Here what is mentioned as target path is "fileset/cgsnapshotpath"

## How risky is this change?
- [ ] Small, isolated change
- [x] Medium, requires regression testing
- [ ] Large, requires functional and regression testing

### CODE REVIEW

**Comment by hemalathagajendran** [comment]
Driver: quay.io/hemalatha_gajendran/cg_serial17:latest
Operator: quay.io/hemalatha_gajendran/cg_threads10:latest

**Comment by hemalathagajendran** [comment]
Image to test: quay.io/hemalatha_gajendran/csi_parallel_snapshot_latest:latest

File: driver/csiplugin/controllerserver.go [hunk 1]
```diff
@@ -208,7 +207,7 @@ func (cs *ScaleControllerServer) createDirectory(ctx context.Context, scVol *sca
 		if scVol.VolPermissions != "" {
 			err = scVol.Connector.MakeDirectoryV2(ctx, scVol.VolBackendFs, targetPath, scVol.VolUid, scVol.VolGid, scVol.VolPermissions)
 			if err != nil {
-				// Directory creation failed, no cleanup will retry in next retry
+				// Directory creation failed, no cleanup will eetry in next retry
```
Comment (deeghuge) [inline driver/csiplugin/controllerserver.go:210]: typo

**Review by deeghuge — CHANGES_REQUESTED** [review]
We will need to changes here 
1. snapshotter timeout = 120s
2. worker-thread = 10

**Comment by hemalathagajendran** [comment]
Images for testing: 
quay.io/hemalatha_gajendran/parallelsnapshot_driver_finalfix_latest:latest
quay.io/hemalatha_gajendran/parallelsnapshot_operator_finalfix:latest

File: driver/csiplugin/controllerserver.go (hunk 1 above)
Comment (hemalathagajendran) [reply driver/csiplugin/controllerserver.go:210]: Done

**Comment by hemalathagajendran** [comment]
> We will need to changes here
> 
> 1. snapshotter timeout = 120s
> 2. worker-thread = 10

Done

**Comment by hemalathagajendran** [comment]
> We will need to changes here
> 
> 1. snapshotter timeout = 120s
> 2. worker-thread = 10

Done

**Review by deeghuge — APPROVED** [review]
LGTM 
do check on number of duplicate request for those 100 snapshot. If those are too many then we have to implement check for snapshot request in progress

I checked for 100 snapshots creation and I could see 117 requests in total for that. 

**root@cluster50-master:~# kubectl logs ibm-spectrum-scale-csi-px8dj -n ibm-spectrum-scale-csi-driver | grep -i 'create snapshot req:' | wc -l
117**

**Logs for duplicate request as the original request completed after timeout of 2 mins:**

root@cluster50-master:~# kubectl logs ibm-spectrum-scale-csi-px8dj -n ibm-spectrum-scale-csi-driver | grep -i 'create snapshot req:' | grep -i 'pvc-42c949fe-0b31-421f-88f2-cc3552f7e0ce'
**I0710 17:30:56.599074**       1 controllerserver.go:2586] [**a8bf14d2-1929-4a4d-8137-391d7af3a899**] CreateSnapshot - create snapshot req: source_volume_id:"1;1;16648342445688148586;0B390B0A:664CD344;03996308-3edf-487c-b369-a205e985ce5a-default;pvc-42c949fe-0b31-421f-88f2-cc3552f7e0ce;/ibm/fs1/03996308-3edf-487c-b369-a205e985ce5a-default/pvc-42c949fe-0b31-421f-88f2-cc3552f7e0ce" name:"**snapshot-24d8e9b7-4aee-407c-a501-ecf4152abe38**"
**I0710 17:33:03.144560**       1 utils.go:90] [**a8bf14d2-1929-4a4d-8137-391d7af3a899**] Time taken to execute /csi.v1.Controller/CreateSnapshot request(in milliseconds): 126545

**I0710 17:45:09.530441**       1 controllerserver.go:2586] [**67f14102-7602-4c33-b0e2-d73b13103587**] CreateSnapshot - create snapshot req: source_volume_id:"1;1;16648342445688148586;0B390B0A:664CD344;03996308-3edf-487c-b369-a205e985ce5a-default;pvc-42c949fe-0b31-421f-88f2-cc3552f7e0ce;/ibm/fs1/03996308-3edf-487c-b369-a205e985ce5a-default/pvc-42c949fe-0b31-421f-88f2-cc3552f7e0ce" name:"**snapshot-24d8e9b7-4aee-407c-a501-ecf4152abe38**"
root@cluster50-master:~#
E0710 17:45:53.765535       1 rest_v2.go:99] [67f14102-7602-4c33-b0e2-d73b13103587] **error in waiting for job completion 1000000064062, [EFSSG0762C The junction path "/ibm/fs1/03996308-3edf-487c-b369-a205e985ce5a-default/snapshot-4bbb711d-bbb3-4a13-9aee-b94dc053b11a/snapshot-24d8e9b7-4aee-407c-a501-ecf4152abe38" already exists.]**
I0710 17:45:53.765557       1 parallelSnapshot.go:117] [67f14102-7602-4c33-b0e2-d73b13103587] Decrease the count of createSnapshotRefLock
I0710 17:45:53.765573       1 parallelSnapshot.go:126] [67f14102-7602-4c33-b0e2-d73b13103587] The target path is unlocked for CreateSnapshot: [03996308-3edf-487c-b369-a205e985ce5a-default/snapshot-4bbb711d-bbb3-4a13-9aee-b94dc053b11a]
**I0710 17:45:53.765596**       1 utils.go:90] [67f14102-7602-4c33-b0e2-d73b13103587] Time taken to execute /csi.v1.Controller/CreateSnapshot request(in milliseconds): 44235

**Review by Jainbrt — APPROVED** [review]

**Pull request merged**